recipe-generator-js-testing/src/IngredientTest.spec.js

```diff
@@ -0,0 +1,252 @@
+            new AllParamsIngredient()
+                .withBooleanArg(true)
+                .withEnumArg(TestEnum.B)
+                .withFlagArg()
+                .withStringArg("foobar")
+                .withIntArg(-10)
+                .withFloatArg(1.123)
+                .withEnumArrayArg([TestEnum.A, TestEnum.B])
+                .withVarargArg("foo", "bar")
+                .withVarargArrayArg([1, 2], [3, 4])
+        );
+    });
+
+    it("should serialize an ingredient with a string default containing quotes", () => {
+        expectJsonEquals(`{"IngredientWithStringDefaultContainingQuotes":{"required":"\\"foo"}}`, new IngredientWithStringDefaultContainingQuotes());
+    });
+
+    it("should serialize an ingredient with a required vararg", () => {
+        expectJsonEquals(`{"IngredientWithRequiredVararg":{"required":["foo","bar"]}}`, new IngredientWithRequiredVararg("foo", "bar"));
+    });
+
+    it("should serialize an ingredient with a required and a required vararg", () => {
+        expectJsonEquals(`{"IngredientWithRequiredAndRequiredVararg":{"required":5,"requiredVararg":[]}}`, new IngredientWithRequiredAndRequiredVararg(5));
+        expectJsonEquals(`{"IngredientWithRequiredAndRequiredVararg":{"required":5,"requiredVararg":["foo"]}}`, new IngredientWithRequiredAndRequiredVararg(5, "foo"));
+    });
+
+    it("should serialize an ingredient with a defaulted required string array", () => {
+        expectJsonEquals(`{"IngredientWithRequiredStringArrayWithDefault":{"required":["foo","bar"]}}`, new IngredientWithRequiredStringArrayWithDefault());
+    });
+
+    it("should serialize an ingredient with a defaulted required vararg of string arrays", () => {
+        expectJsonEquals(`{"IngredientWithRequiredVarargStringArrayWithDefault":{"required":[["foo","bar"],["moo"]]}}`, new IngredientWithRequiredVarargStringArrayWithDefault());
+    });
+
+    it("should serialize an ingredient with a null string value", () => {
+        expectJsonEquals(`{"AllParamsIngredient":{"stringArg":null}}`, new AllParamsIngredient().withStringArg(null));
+    });
+
+    it("should serialize an ingredient with an array of string values with nulls", () => {
+        expectJsonEquals(`{"AllParamsIngredient":{"stringArrayArg":["foo", null, "bar"]}}`, new AllParamsIngredient().withStringArrayArg(["foo", null, "bar"]));
+    });
+
+    it("should serialize a keyed ingredient with a default key", () => {
+        expectJsonEquals(`{"KeyedIngredientWithDefaultKey":{"key":"foo","required":"foo"}}`, new KeyedIngredientWithDefaultKey("foo"));
+    });
+
+    it("should serialize a keyed ingredient with a default key which is a defaulted param", () => {
+        expectJsonEquals(`{"KeyedIngredientWithDefaultKeyParamIsDefaulted":{"key":"foo","required":"foo"}}`, new KeyedIngredientWithDefaultKeyParamIsDefaulted());
+    });
+});
+
+function expectJsonEquals(expected, actual) {
```

Comment:
Why make the `expected` a string instead of JSON object? Wouldn't a JSON object make all the calls to this method nicer?

Reply:
I kind of like it the way it is. It's checking that the ingedient's JSON (which is a string), is what we expect, so it makes sense that `expected` is a string. Backtick strings make it not so bad.